File: sample-conference-videochat/sample-conference-videochat/Components/Settings/VideoSettingsViewController/VideoSettingsViewController.m
//
//  SettingsViewController.m
//  sample-conference-videochat
//
//  Created by Injoit on 21.06.15.
//  Copyright (c) 2015 QuickBlox Team. All rights reserved.
//

#import "VideoSettingsViewController.h"
#import "Settings.h"

#import "SettingCell.h"
#import "SettingSliderCell.h"
#import "SettingSwitchCell.h"

#import "SwitchItemModel.h"
#import "SliderItemModel.h"
#import "SettingsSectionModel.h"

typedef NS_ENUM(NSUInteger, VideoSettingsSectionType) {
    
    VideoSettingsSectionCameraPostion = 0,
    VideoSettingsSectionSupportedFormats = 1,
    VideoSettingsSectionVideoFrameRate = 2,
    VideoSettingsSectionBandwidth = 3
};

@implementation VideoSettingsViewController

- (NSString *)titleForSection:(NSUInteger)section {
    
    switch (section) {
        case VideoSettingsSectionCameraPostion:
            return @"Switch camera position";
        case VideoSettingsSectionSupportedFormats:
            return @"Video formats";
        case VideoSettingsSectionVideoFrameRate:
            return @"Frame rate";
        case VideoSettingsSectionBandwidth:
            return @"Bandwidth";
    }
    
    return nil;
}

- (NSArray *)videoFormatModelsWithCameraPositon:(AVCaptureDevicePosition)cameraPosition {
    //Grab supported formats
    
    NSArray *formats = [QBRTCCameraCapture formatsWithPosition:cameraPosition];
    NSPredicate *formatsPredicate = [NSPredicate predicateWithFormat:@"width <= %i", 1024];
    formats = [formats filteredArrayUsingPredicate:formatsPredicate];
    
    NSMutableArray *videoFormatModels = [NSMutableArray arrayWithCapacity:formats.count];
    for (QBRTCVideoFormat *videoFormat in formats) {
        
        BaseItemModel *videoFormatModel = [[BaseItemModel alloc] init];
        videoFormatModel.title = [NSString stringWithFormat:@"%tux%tu", videoFormat.width, videoFormat.height];
        videoFormatModel.data = videoFormat;
        [videoFormatModels addObject:videoFormatModel];
    }
    
    return videoFormatModels;
}

- (void)didTapBack:(UIButton *)sender {
    [self applySettings];
    [self.navigationController popViewControllerAnimated:YES];
}

- (void)configure {
    Settings *settings = [[Settings alloc] init];
    
    UIBarButtonItem *backButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"chevron"]
                                                                       style:UIBarButtonItemStylePlain
                                                                      target:self
                                                                      action:@selector(didTapBack:)];
    
    self.navigationItem.leftBarButtonItem = backButtonItem;
    backButtonItem.tintColor = UIColor.whiteColor;
    
    //Camera position section
    __weak __typeof(self)weakSelf = self;
    [self addSectionWith:VideoSettingsSectionCameraPostion items:^NSArray *(NSString *sectionTitle) {
        
        //Camera position section
        SwitchItemModel *switchItem = [[SwitchItemModel alloc] init];
        switchItem.title = @"Back Camera";
        
        switchItem.on = (settings.preferredCameraPostion == AVCaptureDevicePositionBack);
        
        return @[switchItem];
    }];
    //Supported video formats section
    [self addSectionWith:VideoSettingsSectionSupportedFormats items:^NSArray *(NSString *sectionTitle) {
        
        AVCaptureDevicePosition position = settings.preferredCameraPostion;
        NSArray *videoFormats = [weakSelf videoFormatModelsWithCameraPositon:position];
        
        NSArray *formats = [QBRTCCameraCapture formatsWithPosition:position];
        NSPredicate *formatsPredicate = [NSPredicate predicateWithFormat:@"width <= %i", 1024];
        formats = [formats filteredArrayUsingPredicate:formatsPredicate];

        //Select index path
        NSUInteger idx = [formats indexOfObject:settings.videoFormat];
        [weakSelf selectSection:VideoSettingsSectionSupportedFormats index:idx];
        
        return videoFormats;
    }];
    //Frame rate
    [self addSectionWith:VideoSettingsSectionVideoFrameRate items:^NSArray *(NSString *sectionTitle) {
        
        SliderItemModel *frameRateSlider = [[SliderItemModel alloc] init];
        frameRateSlider.title = @"30";
        frameRateSlider.minValue = 2;
        frameRateSlider.currentValue = settings.videoFormat.frameRate;
        frameRateSlider.maxValue = 30;
        
        return @[frameRateSlider];
    }];
    //Video bandwidth
    [self addSectionWith:VideoSettingsSectionBandwidth items:^NSArray *(NSString *sectionTitle) {
        
        SliderItemModel *bandwidthSlider = [[SliderItemModel alloc] init];
        bandwidthSlider.title = @"30";
        bandwidthSlider.minValue = 0;
        bandwidthSlider.currentValue = settings.mediaConfiguration.videoBandwidth;
        bandwidthSlider.maxValue = 2000;
        
        return @[bandwidthSlider];
    }];
}

#pragma mark - UITableViewDelegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    switch (indexPath.section) {
            
        case VideoSettingsSectionSupportedFormats: {
            
            [self updateSelectionAtIndexPath:indexPath];
            break;
        }
        default:
            break;
    }
}

#pragma mark - SettingsCellDelegate

- (void)cell:(BaseSettingsCell *)cell didChageModel:(BaseItemModel *)model {
#if !(TARGET_IPHONE_SIMULATOR)
    if ([model isKindOfClass:[SwitchItemModel class]]) {
        [self reloadVideoFormatSectionForPosition:((SwitchItemModel *)model).on ? AVCaptureDevicePositionBack : AVCaptureDevicePositionFront];
    }
#endif
}

#pragma mark - Helpers

- (void)reloadVideoFormatSectionForPosition:(AVCaptureDevicePosition)position {
    
    NSArray *videoFormatModels = [self videoFormatModelsWithCameraPositon:position];
    
    SettingsSectionModel *section = [self sectionWith:VideoSettingsSectionSupportedFormats];
    section.items = videoFormatModels;
    
    NSArray<QBRTCVideoFormat *> *formats = [QBRTCCameraCapture formatsWithPosition:position];
    
    
    NSString *title = [self titleForSection:VideoSettingsSectionSupportedFormats];
    NSIndexPath *oldIdnexPath = self.selectedIndexes[title];
    //Select index path
    
    NSUInteger idx = section.items.count - 1;
    if (idx >= oldIdnexPath.row) {
        
        BaseItemModel *videoFormatModel = section.items[oldIdnexPath.row];
        QBRTCVideoFormat *videoFormat = videoFormatModel.data;
        
        idx = [formats indexOfObject:videoFormat];
    }
    
    [self selectSection:VideoSettingsSectionSupportedFormats index:idx];
    
    NSIndexSet *sectionToReload = [NSIndexSet indexSetWithIndex:VideoSettingsSectionSupportedFormats];
    [self.tableView reloadSections:sectionToReload withRowAnimation:UITableViewRowAnimationFade];
}

- (void)applySettings {
    
    //APPLY SETTINGS
    Settings *settings = [[Settings alloc] init];
    //Preferred camera positon
    SwitchItemModel *cameraPostion = (id)[self modelWithIndex:0 section:VideoSettingsSectionCameraPostion];
    settings.preferredCameraPostion = cameraPostion.on ? AVCaptureDevicePositionBack : AVCaptureDevicePositionFront;
    
    //Frame rate
    SettingsSectionModel *frameRate = [self sectionWith:VideoSettingsSectionVideoFrameRate];
    SliderItemModel *frameRateSlider = frameRate.items.firstObject;
    
    //bandwidth
    SettingsSectionModel *bandwidth = [self sectionWith:VideoSettingsSectionBandwidth];
    SliderItemModel *bandwidthSlider = bandwidth.items.firstObject;
    
    settings.mediaConfiguration.videoBandwidth = bandwidthSlider.currentValue;
    
    //Supported format
#if !(TARGET_IPHONE_SIMULATOR)
    NSIndexPath *supportedFormatIndexPath = [self indexPathAtSection:VideoSettingsSectionSupportedFormats];
    BaseItemModel *format = [self modelWithIndex:supportedFormatIndexPath.row section:supportedFormatIndexPath.section];
    QBRTCVideoFormat *videoFormat = format.data;
    
    settings.videoFormat =
    [QBRTCVideoFormat videoFormatWithWidth:videoFormat.width
                                    height:videoFormat.height
                                 frameRate:frameRateSlider.currentValue
                               pixelFormat:QBRTCPixelFormat420f];
    [self reloadVideoFormatSectionForPosition:settings.preferredCameraPostion];
#endif
    [settings applyConfig];
    [settings saveToDisk];
    
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Settings/VideoSettingsViewController/VideoSettingsViewController.swift
//
//  VideoSettingsViewController.swift
//  sample-conference-videochat-swift
//
//  Created by Vladimir Nybozhinsky on 04.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit
import AVKit
import QuickbloxWebRTC

enum VideoSettingsSectionType: Int {
    case cameraPostion
    case supportedFormats
    case videoFrameRate
    case bandwidth
}

class VideoSettingsViewController: BaseSettingsViewController {
    //MARK: - Overrides
    override func title(forSection section: Int) -> String {
        switch section {
        case VideoSettingsSectionType.cameraPostion.rawValue:
            return "Switch camera position"
        case VideoSettingsSectionType.supportedFormats.rawValue:
            return "Video formats"
        case VideoSettingsSectionType.videoFrameRate.rawValue:
            return "Frame rate"
        case VideoSettingsSectionType.bandwidth.rawValue:
            return "Bandwidth"
        default:
            return ""
        }
    }
    
    //MARK: - Overrides
    override func configure() {
        //Camera position section
        addSection(with: VideoSettingsSectionType.cameraPostion.rawValue, items: { [weak self] sectionTitle in
            //Camera position section
            let switchItem = SwitchItemModel()
            switchItem.title = "Back Camera"
            switchItem.on = self?.settings.preferredCameraPostion == .back
            return [switchItem]
        })
        //Supported video formats section
        addSection(with: VideoSettingsSectionType.supportedFormats.rawValue, items: { [weak self] sectionTitle in
            var videoFormats = [BaseItemModel]()
            guard let position = self?.settings.preferredCameraPostion, let videoFormatsModels = self?.videoFormatModels(withCameraPositon: position)  else {
                return videoFormats
            }
            videoFormats = videoFormatsModels
            let formats = QBRTCCameraCapture.formats(with: position)
            //Select index path
            let idx: Int = (formats as NSArray).index(of: self?.settings.videoFormat as Any)
            self?.selectSection(VideoSettingsSectionType.supportedFormats.rawValue, index: idx)
            return videoFormats
        })
        //Frame rate
        addSection(with: VideoSettingsSectionType.videoFrameRate.rawValue, items: { [weak self] sectionTitle in
            let frameRateSlider = SliderItemModel()
            frameRateSlider.title = "30"
            frameRateSlider.minValue = 2
            frameRateSlider.maxValue = 30
            if let currentValue = self?.settings.videoFormat.frameRate {
                frameRateSlider.currentValue = currentValue
            } else {
                frameRateSlider.currentValue = 2
            }
            return [frameRateSlider]
        })
        //Video bandwidth
        addSection(with: VideoSettingsSectionType.bandwidth.rawValue, items: { [weak self] sectionTitle in
            var currentValue = 30
            let bandwidthSlider = SliderItemModel()
            bandwidthSlider.title = "30"
            bandwidthSlider.minValue = 0
            if let currValue = self?.settings.mediaConfiguration.videoBandwidth {
                currentValue = currValue
            }
            bandwidthSlider.currentValue = UInt(bitPattern: currentValue)
            bandwidthSlider.maxValue = 2000
            return [bandwidthSlider]
        })
    }
    
    // MARK: - UITableViewDelegate
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        switch indexPath.section {
        case Int(VideoSettingsSectionType.supportedFormats.rawValue):
            updateSelection(at: indexPath)
        default:
            break
        }
    }
    
    // MARK: - SettingsCellDelegate
    override func cell(_ cell: BaseSettingsCell?, didChageModel model: BaseItemModel?) {
        guard let model = model as? SwitchItemModel else {
            return }
        reloadVideoFormatSection(for: model.on ? .back : .front)
    }
    
    override func applySettings() {
        //APPLY SETTINGS
        //Preferred camera positon
        guard let cameraPostion = model(with: 0, section: VideoSettingsSectionType.cameraPostion.rawValue) as? SwitchItemModel else {
            return }
        settings.preferredCameraPostion = cameraPostion.on ? .back : .front
        
        //Supported format
        guard let supportedFormatIndexPath = indexPath(atSection: VideoSettingsSectionType.supportedFormats.rawValue),
            let format = model(with: supportedFormatIndexPath.row, section: supportedFormatIndexPath.section),
            let videoFormat = format.data as? QBRTCVideoFormat else {
                return }
        
        //Frame rate
        guard let frameRate = section(with: VideoSettingsSectionType.videoFrameRate.rawValue),
            let frameRateSlider = frameRate.items.first as? SliderItemModel else { return }
        
        //bandwidth
        guard let bandwidth = section(with: VideoSettingsSectionType.bandwidth.rawValue),
            let bandwidthSlider = bandwidth.items.first as? SliderItemModel else {
                return }
        settings.mediaConfiguration.videoBandwidth = Int(bandwidthSlider.currentValue)
        settings.videoFormat = QBRTCVideoFormat.init(width: videoFormat.width,
                                                     height: videoFormat.height,
                                                     frameRate: frameRateSlider.currentValue,
                                                     pixelFormat: QBRTCPixelFormat.format420f)
    }
    
    //MARK: - Internal Methods
    func videoFormatModels(withCameraPositon cameraPosition: AVCaptureDevice.Position) -> [BaseItemModel]? {
        //Grab supported formats
        let formats = QBRTCCameraCapture.formats(with: cameraPosition)
        var videoFormatModels = [BaseItemModel]()
        for videoFormat in formats {
            let videoFormatModel = BaseItemModel()
            videoFormatModel.title = "\(videoFormat.width)x\(videoFormat.height)"
            videoFormatModel.data = videoFormat
            videoFormatModels.append(videoFormatModel)
        }
        return videoFormatModels
    }
    
    // MARK: - Helpers
    func reloadVideoFormatSection(for position: AVCaptureDevice.Position) {
        guard let videoFormatModels = self.videoFormatModels(withCameraPositon: position),
            let section = self.section(with: VideoSettingsSectionType.supportedFormats.rawValue) else {
                return }
        section.items = videoFormatModels
        let formats = QBRTCCameraCapture.formats(with: position)
        let title = self.title(forSection: VideoSettingsSectionType.supportedFormats.rawValue)
        var idx: Int = section.items.count
        if let oldIdnexPath = selectedIndexes[title] {
            //Select index path
            if idx >= oldIdnexPath.row {
                let videoFormatModel = section.items[oldIdnexPath.row]
                let videoFormat = videoFormatModel.data as? QBRTCVideoFormat
                if let videoFormat = videoFormat {
                    idx = (formats as NSArray).index(of: videoFormat)
                }
            }
        }
        selectSection(VideoSettingsSectionType.supportedFormats.rawValue, index: idx)
        let sectionToReload = NSIndexSet(index: Int(VideoSettingsSectionType.supportedFormats.rawValue))
        tableView.reloadSections(sectionToReload as IndexSet, with: .fade)
    }
}
